From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Glitch <glitch.g3431@gmail.com>
Date: Fri, 8 Apr 2022 22:29:50 -0500
Subject: [PATCH] Add support for other Quilt projects and the Quilt maven

Includes Quilt Mappings on Loom, Loom-Quiltflower, and the Quilt maven.

diff --git a/build.gradle b/build.gradle
index 0e1a2857513fe54f976258a89847f65be2d4832a..48cfffba04c18bc3afe41dbdcbb0d0bf1e1d3944 100644
--- a/build.gradle
+++ b/build.gradle
@@ -44,6 +44,11 @@ repositories {
 		name = 'Fabric'
 		url = 'https://maven.fabricmc.net/'
 	}
+	maven {
+		name = "Quilt"
+		url = "https://maven.quiltmc.org/repository/release"
+	}
+	gradlePluginPortal()
 	mavenCentral()
 }
 
@@ -108,6 +113,14 @@ dependencies {
 	// Kapt integration
 	compileOnly('org.jetbrains.kotlin:kotlin-gradle-plugin:1.6.10') // Must match the version included with gradle.
 
+	// Plugins we apply
+	implementation("io.github.juuxel:loom-quiltflower:1.7.3") {
+		exclude group: "net.fabricmc", module: "fabric-loom"
+	}
+	implementation("org.quiltmc:quilt-mappings-on-loom:4.2.0") {
+		exclude group: "net.fabricmc", module: "fabric-loom"
+	}
+
 	// Testing
 	testImplementation(gradleTestKit())
 	testImplementation('org.spockframework:spock-core:2.1-groovy-3.0') {
diff --git a/src/main/java/net/fabricmc/loom/LoomGradlePlugin.java b/src/main/java/net/fabricmc/loom/LoomGradlePlugin.java
index 49fc762edec712bec2f51cadd0dc95d0f7787c90..969db77e44263f58bd0fa40b6e08e0b7be3efd8f 100644
--- a/src/main/java/net/fabricmc/loom/LoomGradlePlugin.java
+++ b/src/main/java/net/fabricmc/loom/LoomGradlePlugin.java
@@ -50,7 +50,7 @@ import net.fabricmc.loom.util.LibraryLocationLogger;
 public class LoomGradlePlugin implements BootstrappedPlugin {
 	public static final Gson GSON = new GsonBuilder().setPrettyPrinting().create();
 	public static final ObjectMapper OBJECT_MAPPER = new ObjectMapper().configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);
-	public static final String LOOM_VERSION = Objects.requireNonNullElse(LoomGradlePlugin.class.getPackage().getImplementationVersion(), "0.0.0+unknown");
+	public static final String LOOM_VERSION = Objects.requireNonNullElse(LoomGradlePlugin.class.getPackage().getImplementationVersion(), "0.12.0+unknown");
 
 	@Override
 	public void apply(PluginAware target) {
@@ -81,5 +81,8 @@ public class LoomGradlePlugin implements BootstrappedPlugin {
 		LoomTasks.registerTasks(project);
 		DecompilerConfiguration.setup(project);
 		IdeaConfiguration.setup(project);
+
+		project.apply(ImmutableMap.of("plugin", "org.quiltmc.quilt-mappings-on-loom"));
+		project.apply(ImmutableMap.of("plugin", "io.github.juuxel.loom-quiltflower"));
 	}
 }
diff --git a/src/main/java/net/fabricmc/loom/LoomRepositoryPlugin.java b/src/main/java/net/fabricmc/loom/LoomRepositoryPlugin.java
index c0da3e8dabbd1a3419ca12a66ffc5f9652ccd1dc..d029ace8a701453b2603b270e810fec97e53c816 100644
--- a/src/main/java/net/fabricmc/loom/LoomRepositoryPlugin.java
+++ b/src/main/java/net/fabricmc/loom/LoomRepositoryPlugin.java
@@ -70,7 +70,10 @@ public class LoomRepositoryPlugin implements Plugin<PluginAware> {
 			repo.setName("Fabric");
 			repo.setUrl(MirrorUtil.getFabricRepository(target));
 		});
-
+		repositories.maven(repo -> {
+			repo.setName("Quilt Releases");
+			repo.setUrl(MirrorUtil.getQuiltRepository(target));
+		});
 		MavenArtifactRepository mojangRepo = repositories.maven(repo -> {
 			repo.setName("Mojang");
 			repo.setUrl(MirrorUtil.getLibrariesBase(target));
diff --git a/src/main/java/net/fabricmc/loom/configuration/accesswidener/decompile/SingleJarDecompileConfiguration.java b/src/main/java/net/fabricmc/loom/configuration/accesswidener/decompile/SingleJarDecompileConfiguration.java
index 9555e1d40dc11488d62608273b888769dc3227e6..99f05da4d42141ca41187a2cae3f517dae060bda 100644
--- a/src/main/java/net/fabricmc/loom/configuration/accesswidener/decompile/SingleJarDecompileConfiguration.java
+++ b/src/main/java/net/fabricmc/loom/configuration/accesswidener/decompile/SingleJarDecompileConfiguration.java
@@ -80,7 +80,7 @@ public class SingleJarDecompileConfiguration extends DecompileConfiguration<Mapp
 			task.setDescription("Decompile minecraft using the default decompiler.");
 			task.setGroup(Constants.TaskGroup.FABRIC);
 
-			task.dependsOn(project.getTasks().named("genSourcesWithCfr"));
+			task.dependsOn(project.getTasks().named("genSourcesWithQuiltflower"));
 		});
 	}
 }
diff --git a/src/main/java/net/fabricmc/loom/util/Constants.java b/src/main/java/net/fabricmc/loom/util/Constants.java
index f5e694ea750ab9b3c67a2946146cd105908ecc8c..61dbece8cb5ee52bb468cf2ecd7d3375937d2f3a 100644
--- a/src/main/java/net/fabricmc/loom/util/Constants.java
+++ b/src/main/java/net/fabricmc/loom/util/Constants.java
@@ -33,6 +33,7 @@ public class Constants {
 	public static final String VERSION_MANIFESTS = "https://piston-meta.mojang.com/mc/game/version_manifest_v2.json";
 	public static final String EXPERIMENTAL_VERSIONS = "https://maven.fabricmc.net/net/minecraft/experimental_versions.json";
 	public static final String FABRIC_REPOSITORY = "https://maven.fabricmc.net/";
+	public static final String QUILT_REPOSITORY = "https://maven.quiltmc.org/repository/release/";
 
 	public static final int ASM_VERSION = Opcodes.ASM9;
 	public static final String MERCURY_SOURCE_VERSION = JavaCore.VERSION_17;
diff --git a/src/main/java/net/fabricmc/loom/util/MirrorUtil.java b/src/main/java/net/fabricmc/loom/util/MirrorUtil.java
index 3350a985f73e403f086a2a988283cbe4644d0cc2..eef1cd42b1e53497012165e1580c6199600f2f04 100644
--- a/src/main/java/net/fabricmc/loom/util/MirrorUtil.java
+++ b/src/main/java/net/fabricmc/loom/util/MirrorUtil.java
@@ -66,4 +66,12 @@ public class MirrorUtil {
 
 		return Constants.FABRIC_REPOSITORY;
 	}
+
+	public static String getQuiltRepository(ExtensionAware aware) {
+		if (aware.getExtensions().getExtraProperties().has("loom_quilt_repository")) {
+			return String.valueOf(aware.getExtensions().getExtraProperties().get("loom_quilt_repository"));
+		}
+
+		return Constants.QUILT_REPOSITORY;
+	}
 }
diff --git a/src/test/groovy/net/fabricmc/loom/test/integration/DecompileTest.groovy b/src/test/groovy/net/fabricmc/loom/test/integration/DecompileTest.groovy
index 9e93097acaffa47a609eefe8d12a6641ab5c32af..b5356cb8231fea27cfdfc81dfaa36c0578c6659a 100644
--- a/src/test/groovy/net/fabricmc/loom/test/integration/DecompileTest.groovy
+++ b/src/test/groovy/net/fabricmc/loom/test/integration/DecompileTest.groovy
@@ -45,6 +45,7 @@ class DecompileTest extends Specification implements GradleProjectTestTrait {
 
 		where:
 			decompiler 		| task								| version
+			'quiltflower'	| "genSourcesWithQuiltflower"		| PRE_RELEASE_GRADLE
 			'fernflower'	| "genSourcesWithFernFlower"		| PRE_RELEASE_GRADLE
 			'cfr' 			| "genSourcesWithCfr"				| PRE_RELEASE_GRADLE
 	}
